p2p: Allow whitelisting manual connections

brunoerg opened the pull request:
Revives #17167. It allows whitelisting manual connections. Fixes #9923

Since there are some PRs/issues around this topic, I'll list some motivations/comments for whitelisting outbound connections from them:
- Speed-up tx relay/mempool sync for testing purposes (my personal motivation for this) - In #26970, theStack pointed out that we whitelist peers to speed up tx relay for fast mempool synchronization, however, since it applies only for inbound connections and considering the topology `node0 <--- node1 <---- node2 <--- ... <-- nodeN`,  if a tx is submitted from any node other than node0, the mempool synchronization can take quite long.
- https://github.com/bitcoin/bitcoin/pull/29058#issuecomment-1865155764 - "Before enabling -v2transport by default (which I'd image may happen after https://github.com/bitcoin/bitcoin/pull/24748) we could consider a way to force manual connections to be only-v1 or even only-v2 (disabling reconnect-with-v1). A possibility could be through a net permission flag, if https://github.com/bitcoin/bitcoin/pull/27114 makes it in."
- https://github.com/bitcoin/bitcoin/pull/17167#issuecomment-1168606032 - "This would allow us to use https://github.com/bitcoin/bitcoin/pull/25355 when making outgoing connections to all nodes, except to whitelisted ones for which we would use our persistent I2P address."
- Force-relay/mempool permissions for a node you intentionally connected to.

luke-jr committed:
net: Move `NetPermissionFlags::Implicit` verification to `AddWhitelistPermissionFlags`

luke-jr committed:
net_processing: Move extra service flag into InitializeNode

luke-jr committed:
Accept "in" and "out" flags to -whitelist to allow whitelisting manual connections

brunoerg commented:
cc: @furszy @theStack

naumenkogs commented:
Concept ACK

stratospher commented on src/init.cpp:
```diff
@@ -520,9 +520,11 @@ void SetupServerArgs(ArgsManager& argsman)
         "CIDR-notated network (e.g. 1.2.3.0/24). Uses the same permissions as "
-        "-whitebind. Can be specified multiple times." , ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+        "-whitebind. "
+        "Additional flags in and out control whether permissions apply to incoming connections and/or outgoing (default: %s). "
```
why do we mention outgoing? (isn't it incoming only)

brunoerg replied:
No, this PR changes it to allow outgoing as well

brunoerg replied:
Just added it to make it clearer

stratospher reviewed (commented):
concept ACK. whitelisting outgoing connections is a useful feature to have! went through the code and it looks good. 

- integration tests in `p2p_permissions.py` would be nice! (It’d be interesting to see the behaviour when different permissions are set in different connection directions.)
- something i had trouble understanding was if we give a trusted peer special permissions, would it matter who initiated the connection?
- micro nit: next time you rebase/update, maybe reword the 1st commit message to use `NetPermissionFlags::Implicit` instead of PF_ISIMPLICIT?

furszy reviewed (commented):
Concept ACK

Could also add a test framework option to whitelist peers. So we don't have to add `-whitelist` flag everywhere.

brunoerg commented:
> Could also add a test framework option to whitelist peers. So we don't have to add -whitelist flag everywhere.

Sounds good! Gonna implement it!

brunoerg committed:
test: add option to speed up tx relay/mempool sync

when `self.noban_tx_relay=True`, the following flag
`-whitelist=noban,in,out@127.0.0.1`is added to `extra_args`
to speed up tx relay/mempool sync.

brunoerg commented:
force-pushed: Added `self.whitelist_peers` in `test_framework` to avoid having to add `-whitelist` flag everywhere.

glozow commented:
linking #9923
perhaps also relevant: #10594, #10051

jonatack reviewed (commented):
ACK 19f3f3a316660890f03d6ddcb4a4230eb2ff5e91 with some suggestions

brunoerg commented on src/net.h:
```diff
@@ -983,7 +985,8 @@ class CConnman
     CNode* ConnectNode(CAddress addrConnect, const char *pszDest, bool fCountFailure, ConnectionType conn_type) EXCLUSIVE_LOCKS_REQUIRED(!m_unused_i2p_sessions_mutex);
-    void AddWhitelistPermissionFlags(NetPermissionFlags& flags, const CNetAddr &addr) const;
+    void AddWhitelistPermissionFlags(NetPermissionFlags& output_flags, const CNetAddr &addr, const std::vector<NetWhitelistPermissions>& ranges) const;
+    static void InitializePermissionFlags(NetPermissionFlags& flags, ServiceFlags& service_flags);
```
Done

brunoerg commented on src/net_permissions.cpp:
```diff
@@ -85,7 +98,7 @@ bool NetWhitebindPermissions::TryParse(const std::string& str, NetWhitebindPermi
     NetPermissionFlags flags;
     size_t offset;
-    if (!TryParsePermissionFlags(str, flags, offset, error)) return false;
+    if (!TryParsePermissionFlags(str, flags, nullptr, offset, error)) return false;
```
Done

brunoerg commented on src/net_permissions.cpp:
```diff
@@ -52,6 +53,14 @@ bool TryParsePermissionFlags(const std::string& str, NetPermissionFlags& output,
+            else if (permission == "in") connection_direction |= ConnectionDirection::In;
+            else if (permission == "out") {
+                if (!output_connection_direction) {
+                    error = _("whitebind is only used for incoming connections");
```
I just changed the error message and adopted the diff (seems so much better to check whether this is a `nullptr`):
```diff
-                if (!output_connection_direction) {
+                if (output_connection_direction == nullptr) {
+                    // Only NetWhitebindPermissions() should pass a nullptr.
```

brunoerg commented on test/functional/test_framework/test_framework.py:
```diff
@@ -495,6 +496,9 @@ def get_bin_from_version(version, bin_name, bin_default):
             extra_confs = [[]] * num_nodes
         if extra_args is None:
             extra_args = [[]] * num_nodes
+        if self.whitelist_peers:
```
Added right before:
```diff
+ # Whitelist peers to speed up tx relay / mempool sync
```

brunoerg commented:
Force-pushed addressing @jonatack's review.

- Moved `InitializePermissionFlags` out of `CConnman`
- Made `TryParsePermissionFlags` static
- Added more test coverage https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1186367052
- Improved documentation

jonatack reviewed (commented):
Approach ACK, modulo comments and https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1186378760 -- it might be better for that commit to be a pure refactoring.  Thanks for addressing the previous feedback.

brunoerg commented on src/net.cpp:
```diff
@@ -393,6 +393,20 @@ CNode* CConnman::FindNode(const std::string& addrName)
     return nullptr;
 }
 
+static void InitializePermissionFlags(NetPermissionFlags& flags, ServiceFlags& service_flags) {
```
Make sense, going to address it.

brunoerg commented on src/init.cpp:
```diff
@@ -434,7 +434,7 @@ void SetupServerArgs(ArgsManager& argsman)
 #endif
     argsman.AddArg("-blockreconstructionextratxn=<n>", strprintf("Extra transactions to keep in memory for compact block reconstructions (default: %u)", DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-blocksonly", strprintf("Whether to reject transactions from network peers. Automatic broadcast and rebroadcast of any transactions from inbound peers is disabled, unless the peer has the 'forcerelay' permission. RPC transactions are not affected. (default: %u)", DEFAULT_BLOCKSONLY), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-blocksonly", strprintf("Whether to reject transactions from network peers. Automatic broadcast and rebroadcast of any transactions from peers is disabled, unless the peer has the 'forcerelay' permission. RPC transactions are not affected. (default: %u)", DEFAULT_BLOCKSONLY), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
I think it was accidental, sorry. I will revert it.

brunoerg commented:
Force pushed addressing:
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1191183121
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1191191757
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1191197012
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1191201370

furszy reviewed (commented):
The `InitializePermissionFlags` function is added twice, at two different locations, at the first and second commits, then one of them gets erased at the third one.
Would be good to cleanup the history so it's only added once.

brunoerg commented:
> The InitializePermissionFlags function is added twice, at two different locations, at the first and second commits, then one of them gets erased at the third one.
Would be good to cleanup the history so it's only added once.

Fixed! Thanks, @furszy

naumenkogs commented:
ACK 1e09c265a9598df3cc39336e3bcccb993dacf3d8

brunoerg commented on test/functional/test_framework/test_framework.py:
```diff
@@ -495,6 +496,10 @@ def get_bin_from_version(version, bin_name, bin_default):
         if extra_args is None:
             extra_args = [[]] * num_nodes
+        # Whitelist peers to speed up tx relay / mempool sync. Don't use it if testing tx relay or timing.
+        if self.whitelist_peers:
```
> Also, the option shouldn't be set for tests that previously didn't have it set

It's false by default and it has been set up for tests that were already using it.

brunoerg replied:
You're right. I am gonna change it for tests that previously had set it for all nodes.

pinheadmz commented on test/functional/test_framework/test_framework.py:
```diff
@@ -495,6 +496,10 @@ def get_bin_from_version(version, bin_name, bin_default):
+        # Whitelist peers to speed up tx relay / mempool sync. Don't use it if testing tx relay or timing.
+        if self.whitelist_peers:
+            for i in range(len(extra_args)):
+                extra_args[i] = extra_args[i] + ["-whitelist=noban,in,out@127.0.0.1"]
```
I'm not sure this is sufficient functional test coverage. With the exception of the new permissions "in" and "out" which are invalid on master, all tests otherwise pass with this commit (1e09c265a9598df3cc39336e3bcccb993dacf3d8) applied.
I was expecting to see a test that makes an outbound connection to a node that misbehaves and doesn't get banned.

Or was the motivation more about the tx "trickle"? In that case I can see it being hard to test without waiting for time to pass?

pinheadmz reviewed (commented):
concept ACK

Ran all tests locally. Built and reviewed code at each commit. Some questions/comments below. Also I notice this is a fairly old feature request that has been through a few PRs. I wonder if you could address Greg's comments on [this older version](https://github.com/bitcoin/bitcoin/pull/10051)

luke-jr reviewed (changes requested):
`-blocksonly` docs need updating

luke-jr commented on src/net.cpp:
```diff
@@ -554,6 +568,11 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
+    NetPermissionFlags permission_flags = NetPermissionFlags::None;
+    ServiceFlags node_services = nLocalServices;
+    AddWhitelistPermissionFlags(permission_flags, addrConnect, vWhitelistedRangeOutgoing);
+    InitializePermissionFlags(permission_flags, node_services);
```
`node_services` has been discarded here, likely as an oversight trying to rebase it past 8d8eeb422e64ccd08bce92eed2ee9cbbf55ffd67.

Best approach I see is to export it outward, ensuring callers need to handle the API change.

brunoerg replied:
Yea, I'm going to change it

brunoerg commented on src/init.cpp:
```diff
@@ -529,9 +529,11 @@ void SetupServerArgs(ArgsManager& argsman)
         "CIDR-notated network (e.g. 1.2.3.0/24). Uses the same permissions as "
-        "-whitebind. Can be specified multiple times." , ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+        "-whitebind. "
+        "Additional flags \"in\" and \"out\" control whether permissions apply to incoming connections and/or outgoing (default: %s). "
+        "Can be specified multiple times.", "incoming only"), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
> Also I think it's a bit unclear where the "additional flags" go, are they part of [permissions@] ?

Yes, see the docs of `-whitelist`:
```cpp
    argsman.AddArg("-whitelist=<[permissions@]IP address or network>", strprintf("Add permission flags to the peers using the given IP address (e.g. 1.2.3.4) or "
        "CIDR-notated network (e.g. 1.2.3.0/24). Uses the same permissions as "
        "-whitebind. "
        "Additional flags \"in\" and \"out\" control whether permissions apply to incoming connections and/or outgoing (default: %s). "
        "Can be specified multiple times.", "incoming only"), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);

    g_wallet_init_interface.AddWalletOptions(argsman);
```

We call `[permissions@]` as permissions flags, so this adds additional ones.

brunoerg committed:
test: add coverage for whitelisting manual connections

brunoerg commented on test/functional/wallet_import_rescan.py:
```diff
@@ -170,7 +170,6 @@ def setup_network(self):
         self.start_nodes(extra_args=[[]] * self.num_nodes)
         self.import_deterministic_coinbase_privkeys()
         self.stop_nodes()
-
```
Fixed

brunoerg commented:
Force-pushed:

1. Renamed `whitelist_peers` to `noban_tx_relay` and set up it `True` for tests that explicity was whitelisting peers with this purpose of speeding up. - https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1274667131
2. Added test coverage for an outbound node that misbehaves and doesn't get banned. - https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1275440729
3. Updated `-blocksonly` documentation.
4. Changed `InitializePermissionFlags` to deal with `https://github.com/bitcoin/bitcoin/commit/8d8eeb422e64ccd08bce92eed2ee9cbbf55ffd67`  -`https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1275643872

luke-jr commented:
1c1bd6cbc6d96fff7f67186aeb57adfe0ee872c4 doesn't build

brunoerg commented:
> https://github.com/bitcoin/bitcoin/commit/1c1bd6cbc6d96fff7f67186aeb57adfe0ee872c4 doesn't build

sorry, fixing!

brunoerg commented:
> https://github.com/bitcoin/bitcoin/commit/1c1bd6cbc6d96fff7f67186aeb57adfe0ee872c4 doesn't build

Force-pushed fixing it

luke-jr commented on src/net.cpp:
```diff
@@ -437,21 +451,22 @@ static CAddress GetBindAddress(const Sock& sock)
     if (pszDest == nullptr) {
         if (IsLocal(addrConnect))
-            return nullptr;
+            return std::make_pair(nullptr, node_services);
```
nit: Dislike scattering this all over the place. `node_services` doesn't even have a meaning with nullptr. Maybe return `std::optional<...>` instead?

brunoerg replied:
Seems more elegant.

brunoerg commented:
Thanks, @luke-jr! Force-pushed changing `CConnman::ConnectNode` to std::optional

pinheadmz commented on src/net.cpp:
```diff
@@ -362,6 +362,20 @@ bool IsLocal(const CService& addr)
+        NetPermissions::AddFlag(flags, NetPermissionFlags::NoBan);
+    }
+
+    if (NetPermissions::HasFlag(flags, NetPermissionFlags::BloomFilter)) {
+        service_flags = static_cast<ServiceFlags>(service_flags | NODE_BLOOM);
+    }
```
I might be misunderstanding something here, but I think this chunk can be moved to `net_processing.cpp` `InitializeNode()` because:
- the `CNode` will already have its `m_permission_flags` set in `ConnectNode()`
- `m_permission_flags` is all that's needed to set `our_services` inside `InitializeNode()`
- then you won't need to use `std::optional<std::pair<CNode*, ServiceFlags>> ` which is, hey, a real banger of a type!

Otherwise I don't see a reason to compute and then pass around those service flags through all those functions before they are used.  (`ConnectNode()->OpenNetworkConnection()->InitializeNode()`)

As a side note, this looks like a lot of work to allow BLOOM for an *outgoing* peer, which must be a really weird edge case anyway right?

brunoerg replied:
> `m_permission_flags` is all that's needed to set `our_services` inside `InitializeNode()`

Are you sure? I think we need to access of `nLocalServices` for it and we need to do it in `CConnman`, doesn't it?

pinheadmz replied:
Hi sorry just getting back to this review. Check out my top two commits here: 

https://github.com/pinheadmz/bitcoin/commits/zip-27114

I'm still running the tests locally and I could be totally missing something, but this was what I was thinking. I don't think `ConnectNode()` needs to return a pair with the service flags, as long as we keep your other changes that pass the *permission flags* along inside the CNode

brunoerg replied:
done!

pinheadmz reviewed (commented):
code review ACK at d54e7dad7513ab4c587456b56d3a838daa151e4d

built and passed all tests

However as noted below I think the approach can be simplified in regards to local services.

brunoerg commented on src/init.cpp:
```diff
@@ -529,9 +529,11 @@ void SetupServerArgs(ArgsManager& argsman)
         "Specify multiple permissions separated by commas (default: download,noban,mempool,relay). Can be specified multiple times.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
 
-    argsman.AddArg("-whitelist=<[permissions@]IP address or network>", "Add permission flags to the peers connecting from the given IP address (e.g. 1.2.3.4) or "
+    argsman.AddArg("-whitelist=<[permissions@]IP address or network>", strprintf("Add permission flags to the peers using the given IP address (e.g. 1.2.3.4) or "
```
You're correct! Gonna change it.

brunoerg replied:
done!

RandyMcMillan commented:
Concept ACK

brunoerg commented:
Rebased and addressed https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1287433864.

pinheadmz commented on src/init.cpp:
```diff
@@ -531,9 +531,11 @@ void SetupServerArgs(ArgsManager& argsman)
         "Specify multiple permissions separated by commas (default: download,noban,mempool,relay). Can be specified multiple times.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
 
-    argsman.AddArg("-whitelist=<[permissions@]IP address or network>", "Add permission flags to the peers connecting from the given IP address (e.g. 1.2.3.4) or "
+    argsman.AddArg("-whitelist=<[permissions@]IP address or network>", "Add permission flags to the peers using the given IP address (e.g. 1.2.3.4) or "
```
optional nit: I think `connecting from` is still better than `using`

brunoerg replied:
perhaps only "from"? "connecting from" seems inbound only.

pinheadmz replied:
"at" ? I dunno

brunoerg replied:
I'll keep it that way for now

pinheadmz commented on src/net.cpp:
```diff
@@ -2910,9 +2915,9 @@ void CConnman::OpenNetworkConnection(const CAddress& addrConnect, bool fCountFai
         return;
 
     CNode* pnode = ConnectNode(addrConnect, pszDest, fCountFailure, conn_type, use_v2transport);
-
     if (!pnode)
         return;
+
```
optional nit: unnecessary whitespace change?

pinheadmz commented on src/net.h:
```diff
@@ -1071,6 +1071,7 @@ class CConnman
         int64_t m_peer_connect_timeout = DEFAULT_PEER_CONNECT_TIMEOUT;
         std::vector<std::string> vSeedNodes;
         std::vector<NetWhitelistPermissions> vWhitelistedRange;
+        std::vector<NetWhitelistPermissions> vWhitelistedRangeOutgoing;
```
Might be nice for future devs looking at this code to rename the original `vWhitelistedRange` to `vWhitelistedRangeIncoming`, unless there's a reason like is it ever not just for incoming connections?

brunoerg replied:
Done

pinheadmz commented on src/net_permissions.cpp:
```diff
@@ -21,9 +21,10 @@ const std::vector<std::string> NET_PERMISSIONS_DOC{
+static bool TryParsePermissionFlags(const std::string& str, NetPermissionFlags& output, ConnectionDirection* output_connection_direction, size_t& readen, bilingual_str& error)
 {
     NetPermissionFlags flags = NetPermissionFlags::None;
+    ConnectionDirection connection_direction = ConnectionDirection::None;
```
Does this mean that `in` and `out` are now required in conf options? If you defaulted to `::In` here would that maintain backwards compatibility?

pinheadmz replied:
Update: oh I see the default now on line 75 below... why not set that here when you initialize the variable?

luke-jr replied:
It's cleaner when/if the default is changed to both.

brunoerg replied:
> why not set that here when you initialize the variable?

I think the reason is the `|=`operator.

pinheadmz commented on test/functional/test_framework/test_framework.py:
```diff
@@ -96,6 +96,7 @@ def __init__(self) -> None:
         """Sets test framework defaults. Do not override this method. Instead, override the set_test_params() method"""
         self.chain: str = 'regtest'
         self.setup_clean_chain: bool = False
+        self.noban_tx_relay: bool = False
```
I really like this clean up commit! Nice feature. I didn't deeply check all the results from this git grep, but there might still be more to do? Maybe from recent test additions on master?

```
--> git grep whitelist *.py | wc -l
55
```

brunoerg replied:
>  there might still be more to do?

Maybe. I'm checking it again, I only modified the tests that explicity points out that is using it to speedup tx relay. so, there are other tests using it for other purposes.

pinheadmz reviewed (commented):
code review ACK at 51618c9c8a87f3a1cd7fcb9b4f7d4da96f182138

I left some more notes below, but recent updates are great. Looking lean and clean

pinheadmz commented:
oh one more thing! probably should add a release note :-)

brunoerg committed:
docs: add release notes for #27114

brunoerg commented:
Force-pushed:

- added release note
- renamed `vWhitelistedRange` to `vWhitelistedRangeIncoming`
- addressed `self.noban_tx_relay` in more tests

brunoerg commented:
CI failure is unrelated.

pinheadmz reviewed (approved):
ACK d69747ab656899d1014483809d22a2c85cb45d78

Built and ran tests locally on macOS/arm. Confirmed changes since last approval were release note, renaming `vWhitelistedRangeIncoming` and applying the new feature to more tests.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK d69747ab656899d1014483809d22a2c85cb45d78
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmU2rDkACgkQ5+KYS2KJ
yTrcqw//U4yGOXwkUE6Wh2ieFnh0Sw1bBPUSPe9V5WobchD47C4G0bR//zfVMgUH
GOmuqT9nbCtCuVJx4OednoXD7JkEt2PBC4NrZ3xV4uvDZQRFNBSSZpr+AK7ylHIi
dn0ayJMKxima0YduqLzrz+FKmwqnNmYFG3Gw7UjM3PZiWxQMsjk7Lk4QWNoPrT+2
zOuITnkssjYd82qdzuI7w1Elb20knSGpgB+8Iu+NPrXvNGq5fRDtAm4gOzMfHVhY
7YYuDzB/id6WwMMAtEA+Y46GXWJCeowg9gJegADecpa5ghUq19ajxxU03Cxe7L0n
zkp8DcvJbCe3TT8BHSdJnQlHAa8JRtLnswi1uj2pjYuj0IhkBq49eJCIbwOHaukz
/g/+QMEZ363IQJaWaydaMACIZhFfxNR9M2QaCZUZN3X/thTlleMl0NyziH0RLVfD
B6eKDQI8yhielZf1UseyoxX03/HcddPYzSF0hZmnHqkKdge1LiUmQ9+vOKfw+z/p
d0xPALxj273Mqkzm9gK6YzaZG+U4/eO7yu92sqyrLnUNSbMHf+jtnt+o4bvq1m90
/bwMhWopX+qpAYuEgELWJPQxg9iL65KdCrGRGTzeHHR4rtqWcoLob14EPyFuK8hX
NBa75IGYUdBVJojjBFsQXEbBgxFPBrjg2LyIaPFcrAC34bW51JM=
=PmQV
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

naumenkogs commented on src/net.h:
```diff
@@ -1058,7 +1058,8 @@ class CConnman
         int64_t m_peer_connect_timeout = DEFAULT_PEER_CONNECT_TIMEOUT;
         std::vector<std::string> vSeedNodes;
-        std::vector<NetWhitelistPermissions> vWhitelistedRange;
+        std::vector<NetWhitelistPermissions> vWhitelistedRangeIncoming;
```
622a829ca1473814928ba32f003480b2382701a2

what's the benefit of having two separate vectors?

brunoerg replied:
You can specify different permissions for each connection direction.

naumenkogs replied:
perhaps two separate lists (or a mapping to bool) could be added when it's needed, and not now that it's unused?

brunoerg replied:
I don't think it's unused because you can specify `-whitelist=noban,in@ipaddr -whitelist=mempool,out@sameipaddr`.

naumenkogs replied:
good point.

naumenkogs replied:
well, actually....you still can track it with a map<addr, bool>, no?

brunoerg replied:
yes, makes sense.

brunoerg replied:
I think so, but we would have to iterate the whole vector everytime we want to get specific permissions (only outbound or inbound)?

naumenkogs replied:
But we don't do this much anyway?
I think the code could be easily modified for this change, you'd only have to pass addr direction to `AddWhitelistPermissionFlags. I won't insist though.

naumenkogs commented on test/functional/interface_rest.py:
```diff
@@ -52,9 +52,7 @@ class RESTTest (BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
         self.extra_args = [["-rest", "-blockfilterindex=1"], []]
-        # whitelist peers to speed up tx relay / mempool sync
```
edbc1ae33b0a17994331de0660b231e7f00a2bda

I like saving a few LOC, but dropping these comments here and there... Presumably they were helpful for the reader to understand why `whitelist`. And `noban_tx_relay` doesn't immediately mean speeding up.

brunoerg replied:
I agree it's helpful, but I am not sure about having it for all occurances. Perhaps we could put in `test_framework`, e.g.:

```diff
diff --git a/test/functional/test_framework/test_framework.py b/test/functional/test_framework/test_framework.py
index 220e51df38..aeb9a2691c 100755
--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -97,6 +97,7 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
         """Sets test framework defaults. Do not override this method. Instead, override the set_test_params() method"""
         self.chain: str = 'regtest'
         self.setup_clean_chain: bool = False
+        # whitelist peers to speed up tx relay / mempool sync
         self.noban_tx_relay: bool = False
         self.nodes: List[TestNode] 
```

naumenkogs replied:
This could yield false positives: sometimes we need just this flags for a different reason than speeding up, right? I also don't think we can reliably expect a test code reader to figure out his answer ("why noban") is in the test framework.

brunoerg replied:
Fair enough, gonna add comments.

brunoerg commented:
Rebased

jonatack reviewed (commented):
WIP review, will finish soon.

brunoerg commented on src/init.cpp:
```diff
@@ -442,7 +442,7 @@ void SetupServerArgs(ArgsManager& argsman)
 #endif
     argsman.AddArg("-blockreconstructionextratxn=<n>", strprintf("Extra transactions to keep in memory for compact block reconstructions (default: %u)", DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-blocksonly", strprintf("Whether to reject transactions from network peers. Automatic broadcast and rebroadcast of any transactions from inbound peers is disabled, unless the peer has the 'forcerelay' permission. RPC transactions are not affected. (default: %u)", DEFAULT_BLOCKSONLY), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-blocksonly", strprintf("Whether to reject transactions from network peers. Automatic broadcast and rebroadcast of any transactions from any peer is disabled, unless it has the 'forcerelay' permission. RPC transactions are not affected. (default: %u)", DEFAULT_BLOCKSONLY), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
Nice suggestion, just addressed it.

brunoerg commented on src/net.cpp:
```diff
@@ -328,6 +328,16 @@ bool IsLocal(const CService& addr)
+        NetPermissions::AddFlag(flags, NetPermissionFlags::Mempool);
+        NetPermissions::AddFlag(flags, NetPermissionFlags::NoBan);
+    }
+}
```
sgtm!

brunoerg commented:
Force-pushed addressing:

https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1397662013
https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1406727898

furszy reviewed (commented):
left two small things. Nothing important.

brunoerg commented on src/net_permissions.cpp:
```diff
@@ -18,6 +19,21 @@ const std::vector<std::string> NET_PERMISSIONS_DOC{
     "addr (responses to GETADDR avoid hitting the cache and contain random records with the most up-to-date info)"
 };
 
+void NetPermissions::Initialize(NetPermissionFlags& flags)
+{
+    if (HasFlag(flags, NetPermissionFlags::Implicit)) {
+        ClearFlag(flags, NetPermissionFlags::Implicit);
+        if (gArgs.GetBoolArg("-whitelistforcerelay", DEFAULT_WHITELISTFORCERELAY)) {
+            AddFlag(flags, NetPermissionFlags::ForceRelay);
+        }
+        if (gArgs.GetBoolArg("-whitelistrelay", DEFAULT_WHITELISTRELAY)) {
+            AddFlag(flags, NetPermissionFlags::Relay);
+        }
```
Sgtm

brunoerg committed:
net: store `-whitelist{force}relay` values in `CConnman`

brunoerg commented on src/net_permissions.cpp:
```diff
@@ -18,6 +19,21 @@ const std::vector<std::string> NET_PERMISSIONS_DOC{
+        }
+        AddFlag(flags, NetPermissionFlags::Mempool);
+        AddFlag(flags, NetPermissionFlags::NoBan);
+    };
```
done

brunoerg commented:
rebased and fixed issue pointed out by CI

pablomartin4btc reviewed (approved):
tACK d0e3c62d414d91f34bf9c0991a58139ae9365221

Tested it with `bitcoin-qt` and I'm wondering if later as a follow-up we could add a feature as a "nice to have" to identify the "whitelisted" peers (and it permissions passed) in the Peers tab on the Node window (we [show](https://bitcoinexplorer.org/rpc-browser?method=getpeerinfo) already the `permissions@` on `RPC` `getpeerinfo` but not sure about other details on `-whitelist` or `-whitebind` and if we want to provide that info via `RPC`). There's already a "related" `GUI` [request](https://github.com/bitcoin-core/gui/issues/649) but more generic to show the `options` passed on the command line. 

`-whitelist=noban@109.201.168.32 -whitelist=noban,out@167.235.110.208
`![Screenshot from 2023-11-29 19-32-54](https://github.com/bitcoin/bitcoin/assets/110166421/280915b5-f5b0-4012-91b0-822310d0f769)

 
`-whitebind=out@1.2.3.4`
![Screenshot from 2023-11-30 17-16-45](https://github.com/bitcoin/bitcoin/assets/110166421/1bcead2c-eabf-4509-82ec-9e53d4a160b7)

As @glozow [pointed](https://github.com/bitcoin/bitcoin/pull/27114#issuecomment-1516052330) out perhaps you can link the issue #9923 related with this change as "fixes #".

brunoerg commented:
> Tested it with bitcoin-qt and I'm wondering if later as a follow-up we could add a feature as a "nice to have" to identify the "whitelisted" peers (and it permissions passed) in the Peers tab on the Node window (we [show](https://bitcoinexplorer.org/rpc-browser?method=getpeerinfo) already the permissions@ on RPC getpeerinfo but not sure about other details on -whitelist or -whitebind and if we want to provide that info via RPC). There's already a "related" GUI https://github.com/bitcoin-core/gui/issues/649 but more generic to show the options passed on the command line.

Sure, will leave it for a follow-up.

brunoerg commented on src/net_permissions.cpp:
```diff
@@ -52,6 +68,15 @@ bool TryParsePermissionFlags(const std::string& str, NetPermissionFlags& output,
+            else if (permission == "out") {
+                if (output_connection_direction == nullptr) {
+                    // Only NetWhitebindPermissions() should pass a nullptr.
+                    error = _("whitebind may only be used for incoming connections (\"out\" was passed)");
```
I don't think so because it is not changing the default behaviour of `-whitebind`.

brunoerg replied:
I don't think so, because we're not changing the default behaviour of `-whitebind`. The additional flags are specified only in `-whitelist` documentation.

brunoerg commented on src/init.cpp:
```diff
@@ -536,9 +536,11 @@ void SetupServerArgs(ArgsManager& argsman)
         "CIDR-notated network (e.g. 1.2.3.0/24). Uses the same permissions as "
-        "-whitebind. Can be specified multiple times." , ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+        "-whitebind. "
+        "Additional flags \"in\" and \"out\" control whether permissions apply to incoming connections and/or outgoing (default: incoming only). "
```
> (default: incoming only - e.g. noban,in,out@1.2.3.4)

I think it might misconfuse. It seems the example (e.g) refers to the default behaviour.

pablomartin4btc replied:
Yeah, I meant to write how to pass the additional flags in the `[permissions@]`'s syntax format (the default behaviour should be without the `out` in it as `noban,in@1.2.3.4`, perhaps that's confusing?). Feel free to ignore if doesn't make sense to you.

brunoerg replied:
I'll leave it as is for now.

brunoerg commented on src/init.cpp:
```diff
@@ -536,9 +536,11 @@ void SetupServerArgs(ArgsManager& argsman)
         "Use [host]:port notation for IPv6. Allowed permissions: " + Join(NET_PERMISSIONS_DOC, ", ") + ". "
         "Specify multiple permissions separated by commas (default: download,noban,mempool,relay). Can be specified multiple times.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Same reason: https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1437020771

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1411359535

luke-jr commented on src/net_permissions.cpp:
```diff
@@ -18,12 +18,28 @@ const std::vector<std::string> NET_PERMISSIONS_DOC{
     "addr (responses to GETADDR avoid hitting the cache and contain random records with the most up-to-date info)"
 };
 
+void NetPermissions::Initialize(NetPermissionFlags& flags, bool whitelist_forcerelay, bool whitelist_relay)
```
Passing options individually like this is ugly - worse than just passing `const ArgsManager&` IMO. If you really want to cache it as a bool, maybe make a struct for connection permission options or something to pass instead.

Another possibility is to just cache a single `NetPermissionFlags` for these.

brunoerg replied:
I think we can avoid passing them in `Initialize` by setting them in `AddWhitelistPermissionFlags`.

pablomartin4btc replied:
I agree to move it to the `CConman::AddWhitelistPermissionFlags` (since the obj has both `whitelist_forcerelay` & `whitelist_relay` vars already set) and you could have something like `CConman:UpdateNetPermissionsFlags`with the logic below.

brunoerg replied:
I will address it in the next push.

brunoerg commented on src/net.cpp:
```diff
@@ -2909,12 +2915,13 @@ void CConnman::OpenNetworkConnection(const CAddress& addrConnect, bool fCountFai
     pnode->grantOutbound = std::move(grant_outbound);
 
-    m_msgproc->InitializeNode(*pnode, nLocalServices);
+    ServiceFlags nodeServices = nLocalServices;
+    m_msgproc->InitializeNode(*pnode, nodeServices);
```
I don't think so because it's expected to not modify `nLocalServices`. Doesn't it?

```cpp
    /**
     * Services this node offers.
     *
     * This data is replicated in each Peer instance we create.
     *
     * This data is not marked const, but after being set it should not
     * change.
     *
     * \sa Peer::our_services
     */
    ServiceFlags nLocalServices;
```

naumenkogs replied:
What does `\sa` mean?

stratospher replied:
a doxygen command :)
https://www.doxygen.nl/manual/commands.html#cmdsa

brunoerg commented on test/functional/p2p_filter.py:
```diff
@@ -93,9 +93,9 @@ def merkleblock_received(self, value):
         self.extra_args = [[
-            '-peerbloomfilters',
-            '-whitelist=noban@127.0.0.1',  # immediate tx relay
+            '-peerbloomfilters'
```
I will address it in the next push.

brunoerg commented on test/functional/p2p_filter.py:
```diff
@@ -93,9 +93,9 @@ def merkleblock_received(self, value):
 class FilterTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 1
+        self.noban_tx_relay = True
```
I don't think so because some tests rely on `-whitelist` for other purposes.

brunoerg commented:
Force-pushed: Rebased and addressed https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1365833858 and https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1440760716 (I moved the `-whitelistrelay` and `-whitelistforcerelay` verification to the `AddWhitelistPermissionFlags` function.

luke-jr commented:
Why was it rebased? Just seems to make it harder to review the changes...

brunoerg commented:
Pushed to fix CI.

luke-jr commented:
>`git range-diff bitcoin-core/master a b` does not care about rebases, see the docs in this repo.

It's also painful to review. If there's no benefit to a rebase, better to just avoid it.

pablomartin4btc reviewed (commented):
> addressed [#27114 (comment)](https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1365833858) and [#27114 (comment)](https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1440760716)

re ACK 6175a2ee09663f7cf20a048b33e537442dfeb81d

(nit: if you have to touch 2nd and/ or 4th commits, naming could be shorter)

naumenkogs commented:
ACK 6175a2ee09663f7cf20a048b33e537442dfeb81d

dergoegge reviewed (commented):
I don't see what this would be useful for, could you elaborate on the motivation?

glozow reviewed (commented):
+1 Could you update the OP with motivation for this PR? I didn't see much from the linked PRs

brunoerg commented:
@dergoegge @glozow I just updated the PR description.

cc: @sr-gi (asked same in IRC)

brunoerg commented on src/init.cpp:
```diff
@@ -567,9 +567,11 @@ void SetupServerArgs(ArgsManager& argsman)
 
-    argsman.AddArg("-whitelist=<[permissions@]IP address or network>", "Add permission flags to the peers connecting from the given IP address (e.g. 1.2.3.4) or "
+    argsman.AddArg("-whitelist=<[permissions@]IP address or network>", "Add permission flags to the peers using the given IP address (e.g. 1.2.3.4) or "
         "CIDR-notated network (e.g. 1.2.3.0/24). Uses the same permissions as "
-        "-whitebind. Can be specified multiple times." , ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+        "-whitebind. "
+        "Additional flags \"in\" and \"out\" control whether permissions apply to incoming connections and/or outgoing (default: incoming only). "
+        "Can be specified multiple times.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
> Why is the approach to add "in" and "out" instead of just having specified permissions apply to both? Are there instances where you want to specify different permissions for the same address depending on the connection direction?

For the same address especifically, idk. But since we can use `-whitelist` to give permissions for "all" nodes (e.g. using `0.0.0.0`), it could be worth to specify the direction.

mzumsande commented:
> If so, it seems more natural to me to specify outbound net permissions wherever that manual connection is instructed

That would mean adding logic / additional options for it in possibly multiple spots, and having yet another place where net permissions are defined. If that becomes too messy, another option would be to leave it in `-whitelist` but just not apply it to automatic outbound connections.

brunoerg commented:
> That would mean adding logic / additional options for it in possibly multiple spots, and having yet another place where net permissions are defined. If that becomes too messy, another option would be to leave it in -whitelist but just not apply it to automatic outbound connections.

I agree. I think the easiest path would be not applying these permissions for the automatic outbound connections, leaving it for the manual ones.

brunoerg commented:
Thanks for the reviews! Since there is a consensus about being useful only for manual connections, I just changed the approach. The flags won't be applied for automatic connections, only for manual ones. I updated the title and description.

brunoerg commented on test/functional/test_framework/test_framework.py:
```diff
@@ -494,6 +495,10 @@ def get_bin_from_version(version, bin_name, bin_default):
         if extra_args is None:
             extra_args = [[]] * num_nodes
+        # Whitelist peers to speed up tx relay / mempool sync. Don't use it if testing tx relay or timing.
+        if self.noban_tx_relay:
```
jsut changed the approach, thanks: https://github.com/bitcoin/bitcoin/pull/27114#issuecomment-1896635863

luke-jr commented:
Concept NACK to restricting it to only manual connections. This will cause random breakage when automatic peering happens to connect to a manual peer first.

mzumsande commented:
> Concept NACK to restricting it to only manual connections. This will cause random breakage when automatic peering happens to connect to a manual peer first.

did you see #28895? We don't make automatic connections to manually specified peers anymore.

luke-jr commented:
Hmm, it's still weird, but I guess it's not terrible then.

sr-gi commented on test/functional/feature_fee_estimation.py:
```diff
@@ -132,11 +132,13 @@ def make_tx(wallet, utxo, feerate):
         self.num_nodes = 3
+        # whitelist peers to speed up tx relay / mempool sync
+        self.noban_tx_relay = True
         # Force fSendTrickle to true (via whitelist.noban)
```
nit: This comment is not relevant anymore. Either replace it with the one right before `self.noban_tx_relay = True` or remove it

brunoerg replied:
Nice catch! I'll address it.

sr-gi reviewed (commented):
Approach ACK [1d0216e](https://github.com/bitcoin/bitcoin/pull/27114/commits/1d0216ed3225056871fee8291e43f24e011d9bdf)

brunoerg commented on src/net.h:
```diff
@@ -1543,6 +1546,18 @@ class CConnman
      */
     std::vector<CService> m_onion_binds;
 
+    /**
+     * flag for adding 'forcerelay' permission to whitelisted inbound
+     * peers with default permissions.
+     */
+    bool whitelist_forcerelay;
+
+    /**
+     * flag for adding 'relay' permission to whitelisted inbound
+     * peers with default permissions.
+     */
+    bool whitelist_relay;
+
```
Yes, I'll address it.

brunoerg commented on src/net_permissions.cpp:
```diff
@@ -18,12 +18,22 @@ const std::vector<std::string> NET_PERMISSIONS_DOC{
+void NetPermissions::Initialize(NetPermissionFlags& flags)
+{
+    if (HasFlag(flags, NetPermissionFlags::Implicit)) {
+        ClearFlag(flags, NetPermissionFlags::Implicit);
```
Where is it calling `NetPermissions::Initialize` again?

sr-gi replied:
Sorry, my bad, looking at the diff I misinterpreted the definition of initialize that is right bellow NetPermissions::ClearFlag as a call to it. Disregard this commet.

brunoerg commented on src/net.h:
```diff
@@ -1058,6 +1054,8 @@ class CConnman
         std::vector<std::string> m_specified_outgoing;
         std::vector<std::string> m_added_nodes;
         bool m_i2p_accept_incoming;
+        bool whitelist_forcerelay = false;
+        bool whitelist_relay = true;
```
Yes, we can call `Init` without setting all values in `Options`. I will change it to use `DEFAULT_WHITELISTFORCERELAY` and `DEFAULT_WHITELISTRELAY`.

brunoerg commented on src/net.cpp:
```diff
@@ -558,9 +564,15 @@ void CNode::CloseSocketDisconnect()
+        }
+    }
+    if (NetPermissions::HasFlag(flags, NetPermissionFlags::Implicit)) {
+        if (whitelist_forcerelay) NetPermissions::AddFlag(flags, NetPermissionFlags::ForceRelay);
+        if (whitelist_relay) NetPermissions::AddFlag(flags, NetPermissionFlags::Relay);
```
I think that was a result of many changes during the path. In fact, we do not need `Initialize` anymore. We can concentrate the logic into `AddWhitelistPermissionFlags`.

sr-gi replied:
Sorry, my bad, looking at the diff I misinterpreted the definition of initialize that is right bellow `NetPermissions::ClearFlag` as a call to it. Disregard this commet.

brunoerg commented on src/net_processing.cpp:
```diff
@@ -1550,6 +1550,11 @@ void PeerManagerImpl::InitializeNode(CNode& node, ServiceFlags our_services)
         m_node_states.emplace_hint(m_node_states.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(node.IsInboundConn()));
         assert(m_txrequest.Count(nodeid) == 0);
     }
+
+    if (NetPermissions::HasFlag(node.m_permission_flags, NetPermissionFlags::BloomFilter)) {
+        our_services = static_cast<ServiceFlags>(our_services | NODE_BLOOM);
+    }
+
```
Because we will need in `OpenNetworkConnection`.

sr-gi replied:
Ohh ok, that used to only be relevant to inbound but we are now also applying it to manual

brunoerg commented on src/net_permissions.cpp:
```diff
@@ -52,6 +62,15 @@ bool TryParsePermissionFlags(const std::string& str, NetPermissionFlags& output,
             else if (permission == "all") NetPermissions::AddFlag(flags, NetPermissionFlags::All);
             else if (permission == "relay") NetPermissions::AddFlag(flags, NetPermissionFlags::Relay);
             else if (permission == "addr") NetPermissions::AddFlag(flags, NetPermissionFlags::Addr);
+            else if (permission == "in") connection_direction |= ConnectionDirection::In;
+            else if (permission == "out") {
```
Nothing should happen. Should we throw an error in this case?

sr-gi replied:
I guess that depends on what feedback we want to provide the user when an invalid configuration has been set (i.e. silently dropping vs being loud about it).

This is quite similar to other misconfigurations, such as setting the same permission twice, but with the particularity that it actually cannot be applied by itself, so maybe it's worth returning an error in this case

brunoerg replied:
Sounds good, I'll address it.

brunoerg replied:
Since we have the `all` flag, I do not see how useful would be allowing passing only conn direction.

brunoerg commented on src/net_permissions.cpp:
```diff
@@ -52,6 +62,15 @@ bool TryParsePermissionFlags(const std::string& str, NetPermissionFlags& output,
+            else if (permission == "in") connection_direction |= ConnectionDirection::In;
+            else if (permission == "out") {
+                if (output_connection_direction == nullptr) {
+                    // Only NetWhitebindPermissions() should pass a nullptr.
```
Sounds good

sr-gi commented on src/init.cpp:
```diff
@@ -567,9 +567,11 @@ void SetupServerArgs(ArgsManager& argsman)
         "CIDR-notated network (e.g. 1.2.3.0/24). Uses the same permissions as "
-        "-whitebind. Can be specified multiple times." , ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+        "-whitebind. "
+        "Additional flags \"in\" and \"out\" control whether permissions apply to incoming connections and/or outgoing (default: incoming only). "
```
nit: and/or manual outgoing

brunoerg replied:
I'll fix it

brunoerg commented:
Force-pushed addressing:
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1480525652
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1480624819
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1481727947
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1483435534

Thanks @sr-gi for reviewing!

sr-gi commented:
tACK [c10f528](https://github.com/bitcoin/bitcoin/pull/27114/commits/c10f528350152ca9248e8c167b67993fc7321ca3)

naumenkogs commented:
ACK c10f528350152ca9248e8c167b67993fc7321ca3

pinheadmz commented on src/net.h:
```diff
@@ -1064,6 +1060,8 @@ class CConnman
         std::vector<std::string> m_specified_outgoing;
         std::vector<std::string> m_added_nodes;
         bool m_i2p_accept_incoming;
+        bool whitelist_forcerelay = DEFAULT_WHITELISTFORCERELAY;
+        bool whitelist_relay = DEFAULT_WHITELISTRELAY;
```
nit, I think these should get the `m_` prefix. When I was reviewing a7240eb99dc792fdf3a6f5aeb93c40c42a8cdc16 I thought `whitelist_relay` was a local variable and couldn't tell where it was defined. Maybe just me though 🤷

pinheadmz commented on doc/release-notes-27114.md:
```diff
@@ -0,0 +1,2 @@
+- Additional flags "in" and "out" have been added to `-whitelist` to control whether
+  permissions apply to incoming connections and/or manual (default: incoming only).
```
Kinda funny that in/out maps to in/manual ... perhaps `"manual"` should've been the permission flag?

brunoerg replied:
I'll leave it as is for while since it is specified that it applies only for manual connections.

pinheadmz reviewed (approved):
ACK c10f528350152ca9248e8c167b67993fc7321ca3

I left comments on a few nits but none are breaking for me. Happy to re-ACK if you update.

Built and ran all functional and unit tests locally on macOS/arm

Reviewed specific changes since my last ACK:
```
git range-diff abfc8c901d..d69747ab65 5883a8911a..c10f528350
```

- Move relay flags to netpermissions from net
- Only apply new logic to manual connections
- Apply flag to new tests added since last review



<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK c10f528350152ca9248e8c167b67993fc7321ca3
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmXVHeAACgkQ5+KYS2KJ
yTqDMxAApyQUw15xZWgrgO+pvNMlRKensYTKTHvPUZK8MeLpM6MfBVruWAnZwxlC
2qj6qyAP+j7mOAOYqZqNo5fgwfD01ruNCL/bCpwAa51/hD3yPf9I/h1nA6AFNoVT
y9JvyknUKWHRWceNi1vBMa9lrkt5IkJ1o/padFaa52OO01OEBTf+ZJQ7FyHlFcQ2
nttow4AIXPM37mYFLil9DGOc90qn728fICC6QrUDXjLL2AGFCPJfysl4vv5FwX6E
92txS6FDh5P8z54DxTOiFSlTVemwOtmSZAtM5Kvs7tgeAiBlOmD9vsMC9BUwRs+c
7Swn9MUiIZ+kveh6oS7SGAQv3uj4kOyhyOPV3iKJKylEaFoivUN2YMtdhNUwUC8i
bz8p5nbkO397HvHuCQ3mEUcWfYeTXS2ozZfRcbvTXqnZMb5QCuM7CPkH+w5QG/hi
yGQDp+SbqDFLiEWejmFysIKhDeXuz110jhydgr77UuiCcTVjPTUCvrEw0hZ/PkZi
ev/tnL0xsC2Nm3gnsjY5ilE+kbkjEerf19qgJ45lweMI2pvT+BNSWxFYxSNbQpYm
ayYcnsYfHbHdoPlLJoTUO44yOyWSdsqu6tLBdYYuH1nDkoijaTl27O9SI5Cxe2E0
9BdBUPu2c9STB1VZzZLNOTiRgKyXq/rYloSyksiI6Jd4HK48zGI=
=rbZH
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

brunoerg commented on src/net.cpp:
```diff
@@ -1789,9 +1796,6 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
     uint64_t nonce = GetDeterministicRandomizer(RANDOMIZER_ID_LOCALHOSTNONCE).Write(id).Finalize();
 
     ServiceFlags nodeServices = nLocalServices;
-    if (NetPermissions::HasFlag(permission_flags, NetPermissionFlags::BloomFilter)) {
-        nodeServices = static_cast<ServiceFlags>(nodeServices | NODE_BLOOM);
-    }
```
Sure, I'll address it.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1496499176. Just removed an unecessary variable.

cc: @naumenkogs @sr-gi

pinheadmz reviewed (approved):
ACK 231e1797023c3502e1605b263b24f29eb659ce55
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 231e1797023c3502e1605b263b24f29eb659ce55
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmXcuNkACgkQ5+KYS2KJ
yTq5oxAAyNHLUhXS4t6pe+5obxZMLEleSnZvMbq7ILrxCeNc0ke7KQKmy/sMFOT9
5pBN8Pc3e/LLtxlx+rNuspiooGqBKz3tRV0fEdAdBt4gQhoyaA7+KuFxfOZXjM5z
TnimlvgAxZk1bI8RWP/Pd/hvd5rMqWBKu5SLtyycVnZ+lQCObtAyMQVart9KjLpw
ORTCUPJgSZjc3OqcFvjyKkxKkZEQgmqd67h4j3+k6AoKRsJwauqPZMALYVtLFPb0
T7xf/6BPk51cyVhflNzv7J3MIOQE4ojlLvPRi+Q3bRZJCoEuPYtfzA8NWz/bKS8o
M4jMuHanvMiLUbBRWWYZ1nv1AlgB9PBcSODi0iU1C3ZBppAb3uG+AYR3xP0NVnBF
F2uheUGrKKsGgimdUFg8sjSTjwEZhAJOq0AwN/D1subNOfJl6jCMo3sF9NnECFBn
plvZoyU3rJSfdfukQOYeF7Q+xF/GxpUUZVSWKtgenpCoc4KvM6c+3aclfDmFKe1s
jgMal6sXkEGjjWIHeCs/fsdMg2PWK8iQqX0iuTDYc1RsD1XjIwiYdgO3oy5oDVjm
M7eDUd04ISsqFVDyY6j/F+jxKFZT1cZxzOj2ACZJegaWUW/xQrVBZ1CD64twtT7p
Ys8svqCF8JiKes0LSWZSG8aHEVz4yx53FukltCljdtbd3CBCM0g=
=yWJM
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

brunoerg commented on src/net_permissions.h:
```diff
@@ -71,6 +76,7 @@ class NetPermissions
         using t = typename std::underlying_type<NetPermissionFlags>::type;
         flags = static_cast<NetPermissionFlags>(static_cast<t>(flags) & ~static_cast<t>(f));
     }
+    static void Initialize(NetPermissionFlags& flags);
```
Done

brunoerg commented:
Force-pushed addressing:
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1505265980
- https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1504102110

thanks, @stratospher

sr-gi reviewed (commented):
re-ACK [0a53361](https://github.com/bitcoin/bitcoin/pull/27114/commits/0a533613fb44207053796fd01a9f4b523a3153d4)

pinheadmz reviewed (approved):
ACK 0a533613fb44207053796fd01a9f4b523a3153d4

confirmed minimal diff since last ack

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 0a533613fb44207053796fd01a9f4b523a3153d4
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmXg8cMACgkQ5+KYS2KJ
yTqb9BAAlsYlKQ6Umtb5FVctkQebcoO7F4P0cHseOl97Xlkcr+slqDGDnYNs19ft
6Ct9HJkB/faLzOQ2gyGq3cOxMevkuQEWBxAKti54uoHY5X5Ye51W5ejJz3BLT+Sw
pDhwlpXo2hSz7GhgpUeb6mj0+rHyA3pMXNvFGjcrjHCX4AeJdWLZEcau0avZ+Luj
5cl0cDxNztu6xWCSAdoooEDyHcpXifYafPOkMV2L6qefh94rZuHuc7gmCnff5geB
lb5DZJQm6lJiLLQBHQ6UUw5hLx4Myfn0Vdb3h937amiwZ+2RsTEHTT1RoZ0lT1Q4
qi/ehtXq+iLcqAA2FzHaiTc1IgZzG3ylJQMGov5rQ2irIpYBKyzPBgZcSloS49oK
lbMB3a3onyABtmmyCQGLG/ituxp8n1/MawyANaQj+uJF7aJrVoylkNctm7J5gb5T
6YYeYVEV/Dav+OqeJj+LZr+4zxvYHndTgrE7MgTZsGMoam/A7jzjZdS3Y0nceCIZ
Lm2d57+nzar/36pp2zzfS6yTjdIgK1NxGEYDLDSBSXdWV3YVaCbGezMzghAc2Wb1
l5JHH+SFvXhOY5WaciuhOUWF7wq726WtaNUeUGldIGHqJB0QrkJOB1IUqpkbaCOU
d4g4ePAXMkxIeo+m9QXwARLX9iFHr4AkW+qsytQoS3OgMQIQpdQ=
=N5cR
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

stratospher reviewed (commented):
concept ACK! it's useful to be able to whitelist and give special permissions to manual peers which are essentially peers you trust.

I've gone through all the commits but not leaving a code review/approach ACK since I don't understand why we'd need different permissions for each connection direction. (related comments - https://github.com/bitcoin/bitcoin/pull/17167#issuecomment-543062796, https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1450191936, https://github.com/bitcoin/bitcoin/pull/27114#discussion_r1374410340)

achow101 commented:
ACK 0a533613fb44207053796fd01a9f4b523a3153d4

sr-gi commented on src/net_permissions.cpp:
```diff
@@ -61,7 +71,16 @@ bool TryParsePermissionFlags(const std::string& str, NetPermissionFlags& output,
+    // By default, whitelist only applies to incoming connections
+    if (connection_direction == ConnectionDirection::None) {
+        connection_direction = ConnectionDirection::In;
+    } else if (flags == NetPermissionFlags::None) {
```
Going over this PR again, shouldn't this have been a separate `if` statement instead of an `if/else` (or not a condition at all)?

Right now we are accepting expressions of the type `@ip:port` (which where already accepted before), but not expressions of the type `in/out@ip:port`.

The former will pass and default to `ConnectionDirection::In +  NetPermissionFlags::Implicit`, while the latter will be rejected. That feels inconsistent
